core/storage/wal.rs

```diff
@@ -194,63 +225,176 @@ impl Wal for WalFile {
         pager: &Pager,
         write_counter: Rc<RefCell<usize>>,
     ) -> Result<CheckpointStatus> {
-        let mut shared = self.shared.write().unwrap();
-        for (page_id, _frames) in shared.frame_cache.iter() {
-            // move page from WAL to database file
-            // TODO(Pere): use splice syscall in linux to do zero-copy file page movements to improve perf
-            let page_id = *page_id as usize;
-            if self.ongoing_checkpoint.contains(&page_id) {
-                continue;
-            }
-
-            let page = pager.read_page(page_id)?;
-            if page.is_locked() {
-                return Ok(CheckpointStatus::IO);
+        'checkpoint_loop: loop {
+            let state = self.ongoing_checkpoint.state;
+            log::debug!("checkpoint(state={:?})", state);
+            match state {
+                CheckpointState::Start => {
+                    // TODO(pere): check what frames are safe to checkpoint between many readers!
+                    self.ongoing_checkpoint.min_frame = self.min_frame;
+                    self.ongoing_checkpoint.max_frame = self.max_frame;
+                    self.ongoing_checkpoint.current_page = 0;
+                    self.ongoing_checkpoint.state = CheckpointState::ReadFrame;
+                }
+                CheckpointState::ReadFrame => {
+                    let shared = self.shared.read().unwrap();
+                    if self.ongoing_checkpoint.current_page as usize >= shared.pages_in_frames.len()
+                    {
+                        self.ongoing_checkpoint.state = CheckpointState::Done;
+                        continue 'checkpoint_loop;
+                    }
+                    let page =
+                        shared.pages_in_frames[self.ongoing_checkpoint.current_page as usize];
+                    let frames = shared
+                        .frame_cache
+                        .get(&page)
+                        .expect("page must be in frame cache if it's in list");
+
+                    for frame in frames.iter().rev() {
```

Comment:
why is this .rev()?

Reply:
you want to checkpoint the oldest frame possible, so I just chose to iterate backwards.

Reply:
aren't the first entries in the vector the oldest?

from `fn append_frame`:

```
            let frames = shared.frame_cache.get_mut(&(page_id as u64));
            match frames {
                Some(frames) => frames.push(frame_id),
                None => {
                    shared.frame_cache.insert(page_id as u64, vec![frame_id]);
                    shared.pages_in_frames.push(page_id as u64);
                }
            }
```

Reply:
Sorry, I meant the youngest, because youngest always override the oldest.

Reply:
ah yeah of course, frames are per page granularity. makes sense 👍

Reply:
in fact it would be great to comment this, and TBH imo it would be quite healthy to add more comments elsewhere too explaining why things are done in a certain way